### PR TITLE
Add LDAP authentication with Elytron LDAP realm

### PR DESCRIPTION
**Pull request by hartimcwildfly**
currently tests and documentation are missing. Just tested it in my dev stage -> works
Some tips how to mock the ldap server in unit tests would be nice :)

### CODE REVIEW

**Review by gsmet — CHANGES_REQUESTED** [review]
Thanks. Just a marker that we will need all this to be squashed.

**Comment by hartimcwildfly** [comment]
Ready for merging. I hope everything is fine :)

**Comment by gsmet** [comment]
@hartimcwildfly could you squash everything into a single commit and rebase? Thanks!

**Comment by hartimcwildfly** [comment]
Rebased and squashed everything (Sorry I did not understand what you meant with squashing in your first comment). Let's see if everything still works.

**Comment by hartimcwildfly** [comment]
@gsmet Could you take a look at the pipelines? Seems like your CI is broken:
##[error]No agent found in pool Azure Pipelines which satisfies the specified demands:
     Agent.Version -gtVersion 2.161.0

**Comment by hartimcwildfly** [comment]
All checks passed. Ready for review/merge.

File: docs/src/main/asciidoc/security-ldap.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/security-ldap.adoc]:
```suggestion
This guide demonstrates how your Quarkus application can use an LDAP server to authenticate and authorize your user identities.
```

File: docs/src/main/asciidoc/security-ldap.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/security-ldap.adoc]:
```suggestion
= Quarkus - Using Security with an LDAP Realm
```

File: docs/src/main/asciidoc/security-ldap.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/security-ldap.adoc]:
```suggestion
which is a https://docs.wildfly.org/17/WildFly_Elytron_Security.html#ldap-security-realm[`wildfly-elytron-realm-ldap`] adapter for Quarkus applications.
```

File: docs/src/main/asciidoc/security-ldap.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/security-ldap.adoc]:
```suggestion
The `elytron-security-ldap` extension requires at least one principal query to authenticate the user and its identity.
```

File: docs/src/main/asciidoc/security-ldap.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/security-ldap.adoc]: I'm surprised we can have only one `attribute-mapping`?

File: docs/src/main/asciidoc/security-ldap.adoc
Comment (gsmet) [inline docs/src/main/asciidoc/security-ldap.adoc]:
```suggestion
The application is now protected and the identities are provided by our LDAP server.
```

File: extensions/elytron-security-ldap/deployment/src/test/java/io/quarkus/elytron/security/ldap/LdapSecurityRealmTest.java
Comment (gsmet) [inline extensions/elytron-security-ldap/deployment/src/test/java/io/quarkus/elytron/security/ldap/LdapSecurityRealmTest.java]: I wonder if this should be implemented as a test resource instead? See what we did for H2 for instance (`H2DatabaseTestResource`). For now, the test resource could be in this artifact, no need to create a test artifact.

File: extensions/elytron-security-ldap/runtime/src/main/java/io/quarkus/elytron/security/ldap/config/AttributeMappingConfig.java
Comment (gsmet) [inline extensions/elytron-security-ldap/runtime/src/main/java/io/quarkus/elytron/security/ldap/config/AttributeMappingConfig.java]: You can remove all the names if the property has the right name.

File: extensions/elytron-security-ldap/runtime/src/main/resources/META-INF/quarkus-extension.yaml
Comment (gsmet) [inline extensions/elytron-security-ldap/runtime/src/main/resources/META-INF/quarkus-extension.yaml]: Maybe make it `preview` for some time?

**Review by gsmet — DISMISSED** [review]
I added a few questions and suggestions.

I'm not the security expert here so we will need an approval from @sberyozkin too.

File: extensions/elytron-security-ldap/runtime/src/main/java/io/quarkus/elytron/security/ldap/config/AttributeMappingConfig.java
Comment (hartimcwildfly) [reply extensions/elytron-security-ldap/runtime/src/main/java/io/quarkus/elytron/security/ldap/config/AttributeMappingConfig.java]: Ok, I'm going to fix this.

File: extensions/elytron-security-ldap/deployment/src/test/java/io/quarkus/elytron/security/ldap/LdapSecurityRealmTest.java
Comment (hartimcwildfly) [reply extensions/elytron-security-ldap/deployment/src/test/java/io/quarkus/elytron/security/ldap/LdapSecurityRealmTest.java]: I did this for the integration tests, but I did not test it with the unit tests.
I'm going to take another look and compact it. The ITs and unit tests are similar but I oriented on the elytron jdbc extension which also has overlapping tests.

File: docs/src/main/asciidoc/security-ldap.adoc
Comment (hartimcwildfly) [reply docs/src/main/asciidoc/security-ldap.adoc]: Hmm you are right, one should be able to define multiple attribute-mappings. But currently the quarkus config only allows Map<String, AttributeMappingConfig> but not Map<Integer, AttributeMappingConfig> or List<AttributeMappingConfig>. For my usecase a single attribute-mapping was enough.

File: extensions/elytron-security-ldap/runtime/src/main/resources/META-INF/quarkus-extension.yaml
Comment (hartimcwildfly) [reply extensions/elytron-security-ldap/runtime/src/main/resources/META-INF/quarkus-extension.yaml]: ok

File: docs/src/main/asciidoc/security-ldap.adoc
Comment (hartimcwildfly) [reply docs/src/main/asciidoc/security-ldap.adoc]: "an [...] adapter" is correct.

File: docs/src/main/asciidoc/security-ldap.adoc
Comment (hartimcwildfly) [reply docs/src/main/asciidoc/security-ldap.adoc]: This is semantically wrong. I'm going to fix this, too. (PrincipalQueries are from JDBC Realm)

File: extensions/elytron-security-ldap/deployment/src/test/java/io/quarkus/elytron/security/ldap/LdapSecurityRealmTest.java
Comment (hartimcwildfly) [reply extensions/elytron-security-ldap/deployment/src/test/java/io/quarkus/elytron/security/ldap/LdapSecurityRealmTest.java]: Outsourced this in an extra TestResource. Pushing soon.

File: docs/src/main/asciidoc/security-ldap.adoc
Comment (hartimcwildfly) [reply docs/src/main/asciidoc/security-ldap.adoc]: Added possibility to define multiple attribute-mappings. Pushing soon.

**Comment by hartimcwildfly** [comment]
Ready for another review. Should I squash it again?

**Comment by hartimcwildfly** [comment]
Deleted a duplicate file and squashed everything again.

**Comment by hartimcwildfly** [comment]
Deleted a unused maven dependency. Now ready for merge.
@sberyozkin and @gsmet  please review.

File: extensions/elytron-security-ldap/runtime/src/main/java/io/quarkus/elytron/security/ldap/LdapRecorder.java
Comment (loicmathieu) [inline extensions/elytron-security-ldap/runtime/src/main/java/io/quarkus/elytron/security/ldap/LdapRecorder.java]: This will trigger the creation of a `WildFlyElytronProvider` object on each requests.

Please use a constant to hold the provider.

See PR #6358 that fixes the same issue on the existing security extensions.

**Review by loicmathieu — CHANGES_REQUESTED** [review]
I only review this PR for a known performance issue, please fix the performance issue as requested.

File: extensions/elytron-security-ldap/runtime/src/main/java/io/quarkus/elytron/security/ldap/LdapRecorder.java
Comment (hartimcwildfly) [reply extensions/elytron-security-ldap/runtime/src/main/java/io/quarkus/elytron/security/ldap/LdapRecorder.java]: Fixed.

File: extensions/elytron-security-ldap/runtime/src/main/java/io/quarkus/elytron/security/ldap/LdapRecorder.java
Comment (loicmathieu) [inline extensions/elytron-security-ldap/runtime/src/main/java/io/quarkus/elytron/security/ldap/LdapRecorder.java]: We try to avoid lambda inside Quarkus extension code

**Comment by loicmathieu** [comment]
@hartimcwildfly thanks for the performance fix I asked, it's OK for me. I added a small comment on the usage of lambda.

File: extensions/elytron-security-ldap/runtime/src/main/java/io/quarkus/elytron/security/ldap/LdapRecorder.java
Comment (hartimcwildfly) [reply extensions/elytron-security-ldap/runtime/src/main/java/io/quarkus/elytron/security/ldap/LdapRecorder.java]: Reverted to anonymous function instead of lambda.

**Review by loicmathieu — APPROVED** [review]
LGTM, but only for the perf issue I mention, don't count me as a reviewer.

**Comment by gsmet** [comment]
Current status is: we are waiting for @sberyozkin to review that one.

**Comment by sberyozkin** [comment]
@hartimcwildfly @gsmet sorry for a delay, I was away. I'm catching up and will try to comment soon, thanks

**Comment by sberyozkin** [comment]
@hartimcwildfly @gsmet While I'm not an LDAP expert, this PR does look quite perfect to me, as it is neatly done and simply to follow.  Let me approve. Hi Stuart @stuartwdouglas, have a look please too if you can find some time, I'd be happy to merge tomorrow my time.

**Review by sberyozkin — APPROVED** [review]

File: bom/runtime/pom.xml
Comment (stuartwdouglas) [inline bom/runtime/pom.xml]: It looks like this is only used for testing? If so it should be in the build parent rather than the bom.

**Review by stuartwdouglas — DISMISSED** [review]

File: bom/runtime/pom.xml
Comment (hartimcwildfly) [reply bom/runtime/pom.xml]: Makes sense. Fixed.

**Pull request merged**
